src/osd/scrubber/scrub_backend.h

```diff
@@ -0,0 +1,402 @@
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep,
+               const std::set<pg_shard_t>& acting);
+
+  // Replica constructor: no primary map
+  ScrubBackend(PgScrubber& scrubber,
+               PGBackend& backend,
+               PG& pg,
+               pg_shard_t i_am,
+               bool repair,
+               scrub_level_t shallow_or_deep);
+
+  friend class PgScrubber;
+
+  /**
+   * reset the per-chunk data structure (scrub_chunk_t).
+   * Create an empty scrub-map for this shard, and place it
+   * in the appropriate entry in 'received_maps'.
+   *
+   * @returns a pointer to the newly created ScrubMap.
+   */
+  ScrubMap* new_chunk();
+
+  /**
+   * sets Backend's m_repair flag (setting m_mode_desc to a corresponding
+   * string)
+   */
+  void update_repair_status(bool should_repair);
+
+  void replica_clean_meta(ScrubMap& smap,
+                          bool max_reached,
+                          const hobject_t& start);
+
+  /**
+   * decode the arriving MOSDRepScrubMap message, placing the replica's
+   * scrub-map into received_maps[from].
+   *
+   * @param from replica
+   */
+  void decode_received_map(pg_shard_t from, const MOSDRepScrubMap& msg);
+
+  void scrub_compare_maps(bool max_reached);
+
+  int scrub_process_inconsistent();
+
+  void repair_oinfo_oid(ScrubMap& smap);
+
+  const omap_stat_t& this_scrub_omapstats() const { return m_omap_stats; }
+
+  std::ostream& logger_prefix(std::ostream* _dout, const ScrubBackend* t);
+
+ private:
+  // set/constructed at the ctor():
+  PgScrubber& m_scrubber;
+  PGBackend& m_pgbe;
+  PG& m_pg;
+  const pg_shard_t m_pg_whoami;
+  bool m_repair;
+  const scrub_level_t m_depth;
+  const spg_t m_pg_id;
+  std::vector<pg_shard_t> m_acting_but_me;  // primary only
+  bool m_is_replicated{true};
+  std::string_view m_mode_desc;
+  std::string m_formatted_id;
+ /// collecting some scrub-session-wide omap stats
+  omap_stat_t m_omap_stats;
+
+  // shorthands:
+  ConfigProxy& m_conf;
+  LogChannelRef clog;
+
+ private:
+  using auth_and_obj_errs_t =
+    std::tuple<std::list<pg_shard_t>,  ///< the auth-list
+               std::set<pg_shard_t>    ///< object_errors
+               >;
+
+  std::optional<scrub_chunk_t> this_chunk;
+
+  /// Maps from objects with errors to missing peers
+  HobjToShardSetMapping m_missing;  // used by scrub_process_inconsistent()
+
+  /// Maps from objects with errors to inconsistent peers
+  HobjToShardSetMapping m_inconsistent;  // used by scrub_process_inconsistent()
+
+  /// Cleaned std::map pending snap metadata scrub
+  ScrubMap m_cleaned_meta_map{};
+
+  /// a reference to the primary map
+  ScrubMap& my_map();
+
+  /**
+   *  merge_to_master_set() updates
+   *   - this_chunk->maps[from] with the replicas' scrub-maps;
+   *   - this_chunk->master_set as a union of all the maps' objects;
+   */
+  void merge_to_master_set();
```

Comment:
merge_to_authoritative_set (and comments above)

Reply:
done